File: 0255_Problem_1.hs
-- Problem Link: https://leetcode.com/contest/weekly-contest-255/problems/find-greatest-common-divisor-of-array/

import Control.Monad (liftM2)

-- Original solution featured in the video:
findGCD = liftM2 gcd minimum maximum

-- Note that using `liftA2` (from Control.Applicative) instead of `liftM2` also works and is usually more efficient, as for some types of inputs it can be paralellized.

-- A solution without using `liftM2` or `LiftA2` (to show what is going on under the hood):
findGCD'' nums = gcd (minimum nums) (maximum nums)

File: 0255_Problem_1.rb
# Problem Link: https://leetcode.com/contest/weekly-contest-255/problems/find-greatest-common-divisor-of-array/

# Solution shown in video:
def find_gcd(nums)
    return nums.max.gcd(nums.min)
end

# More idiomatic solution:
#   (Courtesy of YouTube user Luj8n, 
#   c.f. https://www.youtube.com/watch?v=UVUjnzpQKUo&lc=UgzQOYQEEKckxNfgapR4AaABAg )
def find_gcd2(nums)
  nums.minmax.reduce(:gcd)
end

